scene/2d/navigation_agent_2d.cpp

```diff
@@ -96,16 +96,30 @@ void NavigationAgent2D::_bind_methods() {
 
 void NavigationAgent2D::_notification(int p_what) {
 	switch (p_what) {
-		case NOTIFICATION_READY: {
-			agent_parent = Object::cast_to<Node2D>(get_parent());
-			if (agent_parent != nullptr) {
-				// place agent on navigation map first or else the RVO agent callback creation fails silently later
-				NavigationServer2D::get_singleton()->agent_set_map(get_rid(), agent_parent->get_world_2d()->get_navigation_map());
-				set_avoidance_enabled(avoidance_enabled);
-			}
+		case NOTIFICATION_POST_ENTER_TREE: {
+			// need to use POST_ENTER_TREE cause with normal ENTER_TREE not all required Nodes are ready.
+			// cannot use READY as ready does not get called if Node is readded to SceneTree
+			set_agent_parent(get_parent());
 			set_physics_process_internal(true);
 		} break;
 
+		case NOTIFICATION_PARENTED: {
+			if (is_inside_tree() && (get_parent() != agent_parent)) {
+				// only react to PARENTED notifications when already inside_tree and parent changed, e.g. users switch nodes around
+				// PARENTED notification fires also when Node is added in scripts to a parent
+				// this would spam transforms fails and world fails while Node is outside SceneTree
+				// when node gets reparented when joining the tree POST_ENTER_TREE takes care of this
+				set_agent_parent(get_parent());
+				set_physics_process_internal(true);
```

Comment:
@smix8 Does this code ever get called? It looks like `NOTIFICATION_PARENTED` always gets called before the node has its tree assigned:
https://github.com/godotengine/godot/blob/0c815022178ca23fa55085981384caa749f7f18a/scene/main/node.cpp#L1130-L1134

and as far as I know there's no way to reparent a node without removing it from the tree first

Reply:
I dont remember the details but there was a specific issue where a user managed to corrupt the agent connection with the wrong parent due to shuffling node parents with subscenes and this fixed it. Thats also why I left such a large comment so I wouldn't remove it myself in a few months because I would also assume it wouldn't ever be triggered.